plugins/Kaleidoscope-LayerNames/src/kaleidoscope/plugin/LayerNames.cpp

```diff
@@ -0,0 +1,94 @@
+/* Kaleidoscope - Firmware for computer input devices
+ * Copyright (C) 2022  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "kaleidoscope/plugin/LayerNames.h"
+
+#include <Arduino.h>                   // for delay, PSTR, strcmp_P, F, __FlashStri...
+#include <Kaleidoscope-FocusSerial.h>  // for Focus, FocusSerial
+
+#include "kaleidoscope/Runtime.h"                 // for Runtime, Runtime_
+#include "kaleidoscope/plugin/EEPROM-Settings.h"  // for EEPROMSettings
+
+namespace kaleidoscope {
+namespace plugin {
+
+// =============================================================================
+
+EventHandlerResult LayerNames::onNameQuery() {
+  return ::Focus.sendName(F("LayerNames"));
+}
+
+EventHandlerResult LayerNames::onFocusEvent(const char *command) {
+  if (::Focus.handleHelp(command, PSTR("layernames")))
+    return EventHandlerResult::OK;
+
+  if (strcmp_P(command, PSTR("layernames")) != 0)
+    return EventHandlerResult::OK;
+
+  if (::Focus.isEOL()) {
+    uint16_t pos = 0;
+    while (pos < storage_size_) {
+      uint8_t name_size = Runtime.storage().read(storage_base_ + pos++);
+
+      if (name_size == 0 || name_size == 255) break;
+
+      ::Focus.send(name_size);
+
+      for (uint8_t i = 0; i < name_size; i++) {
+        uint8_t b = Runtime.storage().read(storage_base_ + pos++);
+        ::Focus.sendRaw(static_cast<char>(b));
+      }
+      ::Focus.sendRaw(PSTR("\n"));
+    }
+    ::Focus.send(0, storage_size_);
+  } else {
+    uint16_t pos = 0;
+
+    while (pos < storage_size_) {
+      uint8_t name_size;
+      ::Focus.read(name_size);
+
+      // size is followed by a space, ignore that.
+      char spc;
+      ::Focus.read(spc);
+
+      Runtime.storage().update(storage_base_ + pos++, name_size);
+
+      if (name_size == 0 || name_size == 255) break;
+
+      for (uint8_t i = 0; i < name_size; i++) {
+        char c;
+        ::Focus.read(c);
+
+        Runtime.storage().update(storage_base_ + pos++, c);
+      }
+    }
+    Runtime.storage().commit();
+  }
+
+  return EventHandlerResult::EVENT_CONSUMED;
+}
+
+// public
+void LayerNames::reserve_storage(uint16_t size) {
```

Comment:
Is it our standard design to not set a default?

Reply:
Yup, we're not setting defaults of `reserve_storage()` and similar methods, because in most cases what would be a sane default largely depends on the rest of the sketch. ~64 bytes for LayerNames is fine for a firmware with like 5 layers, 12 chars / layer is quite reasonable. But when we have 10 layers like on the default Atreus firmware, 6 bytes is starting to feel limiting. On the other hand, we don't want to allocate ~100 bytes on firmware that has only 5 layers, because 20 chars / layer is a bit over the top.

Defaults would also depend on the hardware: on AVR where EEPROM is **much** more limited, defaults should be conservative. On GD32, where EEPROM is plenty, defaults should be comfortably large.

So I'd rather defer setting a reasonable default to a) documentation; b) sketches shipped in Chrysalis-Firmware-Bundle.